doc/en/2.config.md
[<p align="right">->Back to directory</p>](0.directory.md)


# owlcache Configuration file
* When the program starts, the parameters you dynamically configure will overwrite the parameters in the *.conf file

## Profile writing convention
>1.Indicates that this line is a comment at the beginning of #    
>2.Grouping information with [] (more intuitive display of configuration information)   
>3.The configuration parameter name and parameter value are separated by a space.     

## [Owlcache]
### Set the IP that the program listens
* If you cannot access the service after normal startup, please check whether the firewall or other software restricts your access.
>Host 0.0.0.0

### Set program response IP
* Default local
* Use it in the response result Json string returned by TCP and HTTP, pay attention to the correctness of the IP during actual deployment in the local area network and external network environment!
>ResponseHost 127.0.0.1

### Log file path
* The default log will be generated to the service running directory
* !!! If the Logfile value is specified when starting the service on the command line, and the configuration file is not commented out here, the configuration file shall prevail
>Logfile ./log_file/

### Database file path
* The default database file will be generated to the service running directory
>DBfile ./db_file/


## [TCP]
### TCP listening port
* The default is 7720
* If it is not filled in or the port value is 0, port 7720 will be monitored by default
>Tcpport 7720

### Whether to open the Tcp service
* No password authentication in TCP mode (for production environment deployment, please shield TCP access ports according to the actual situation, etc., pay attention to use safety)
* The values are "1" (open service) and "0" (close service). The default is 1 to enable the service.
>CloseTcp 1


## [HTTP]
### HTTP listening port
* The default is 7721.
* If it is not filled in or the port value is 0, port 7721 will be monitored by default.
>Httpport 7721

### HTTP password
* Mandatory and necessary, don’t set too simple
* Please set your password. The length of the password must not be less than ten digits, and it cannot be a pure number. It is recommended to use uppercase letters + lowercase letters + numbers + simple symbols
>Pass 

### Tonken expiration time
* Set an expiration time for the Tonken value generated by the "Pass" command.
* The default is 0. The unit is seconds. If the value is 0, it will never expire.
* Warning: The setting never expires, there may be a security risk.
>Tonken_expire_time 0

### Set the key of Tonken value
* It is recommended to change the default value.
>Tonken_jwt_secret hello_owlcache

### HTTPS related configuration
* Open_Https Whether to open HTTPS. Values are 0 (closed), 1 (open). Off by default.
* Https_CertFile Cert file path. For example: /www/server.crt, D:/key/server.crt.
* Https_KeyFile Key file path. For example: /www/server.key, D:/key/server.key.
* HttpsClient_InsecureSkipVerify When the HTTPS mode is enabled, whether to verify the validity of the certificate when owlcache communicates with each other. Values are 0 (closed), 1 (open). Off by default. The validity of the certificate will not be verified when it is turned on.
>Open_Https 0

>Https_CertFile /www/server.crt

>Https_KeyFile /www/server.key

>HttpsClient_InsecureSkipVerify 0

### Cross-domain related configuration
* "Cross-origin resource sharing" configuration.
* Cors values are "1" (open service) and "0" (close service). The default value is 0 to close the service and does not allow cross-domain.
* The Access_Control_Allow_Origin value defaults to "*" (allow all domain requests). You can set this value to the domain you specify.
* Note: Be cautious of CSRF attacks.
>Cors 0

>Access_Control_Allow_Origin *


## [HttpGroupData]
### Request timeout time when the clusters communicate with each other
* The unit is milliseconds. The default timeout is 2700 milliseconds. In fact, the data intercommunication time of the intranet environment is very low, and it can be set according to the actual situation.
>HttpClientRequestTimeout 2700

### Http client request settings
* The requested data will be cached locally for a short time to avoid repeated frantically querying other nodes for the same Key in a high concurrency state. Appropriate delayed query helps optimize performance and efficiency.
* The unit is milliseconds, and the default lifetime of this cache is 5000 milliseconds. A value of 0 means no caching (suitable for scenarios with small concurrency and high data real-time requirements).
>HttpClientRequestLocalCacheLifeTime 5000


## [GroupDataSync]

### Whether to enable cluster data synchronization
* 0 means not open. 1 means on. It is not turned on by default.
>GroupDataSync 0

### Set Gossip listening port
* This item will take effect only when the cluster data synchronization service is enabled.
* The cluster data synchronization service relies on the Gossip protocol. It is recommended not to set this item, the system will automatically allocate ports. The default value is 0 (the system automatically monitors a port and outputs the port at startup).
>Gossipport 0

### The entire Gossip cluster needs a unified password
* This item will take effect only when the cluster data synchronization service is enabled.
* A token used to exchange data in the cluster through the Gossip protocol. The default is blank, you need to set the same password for the cluster.
>GossipDataSyncAuthKey 

### Request timeout time when cluster synchronization data communicate with each other
* The unit is milliseconds. The default timeout is 5000 milliseconds. In fact, the data intercommunication time of the intranet environment is very low, and it can be set according to the actual situation.
>GossipHttpClientRequestTimeout 5000


## [Memcache_Client]
### Turn on importing data from memcache
* Whether to enable query for non-existent data, query from memcache and save it in the local database. 0 means not open.
>Get_data_from_memcache 0

### List of memcache queried
* The "|" symbol interval is required between different nodes. For example: 127.0.0.1:11211|192.168.0.77:11211. The default value is empty.
>Memcache_list 

### Expiration time of imported data
* Set an expiration time for the key stored in the local database from memcache. The default is 0, which will never expire. The unit is seconds.
>Get_memcache_data_set_expire_time 0

### MemcacheClient client request settings
* When Max_Error_Number exceeds the limit value, MemcacheClient requests to "pause" the Sleeptime value to optimize the response speed of the program.
* Sleeptime represents sleep time, in seconds, the default minimum sleep is 2 seconds.
* Max_Error_Number represents the maximum number of failed requests, the unit is an integer, and the default minimum error request is 2 times.
>MemcacheClient_Request_Timeout_Sleeptime 2

>MemcacheClient_Request_Max_Error_Number 2


## [Redis_Client]
### Turn on importing data from Redis
* Whether to open the query for non-existent data, query from redis and save it in the local database. 0 means not open.
>Get_data_from_redis 0

### The Redis address to be queried
* Cannot be a null value.  
>Redis_Addr 

### Redis password to be queried
* Cannot be a null value.  
>Redis_Password 

### Need to query Redis database
* Default is 0
>Redis_DB 0

### Expiration time of imported data
* Set an expiration time for the key stored in the local database from Redis. The default is 0, which will never expire. The unit is seconds.
>Get_redis_data_set_expire_time 0

### RedisClient client requests settings
* The RedisClient client requests settings. When Max_Error_Number exceeds the limit value, RedisClient requests to "pause" the Sleeptime value to optimize the response speed of the program.
* Sleeptime represents sleep time, in seconds, the default minimum sleep is 2 seconds.
* Max_Error_Number represents the maximum number of failed requests, the unit is an integer, and the default minimum error request is 2 times.
>RedisClient_Request_Timeout_Sleeptime 2

>RedisClient_Request_Max_Error_Number 2


## [Url_Cache]
### Enable importing data from URL address
* Cache the HTTP response data obtained after URL proxy access to Owlcache
* The values are "1" (open service) and "0" (close service). The default is 0 to turn off the service.
>Open_Urlcache 0

### The name of the configuration file that needs to be loaded after URL caching is enabled
* The format is XML, it exists in the DBfile configuration item directory, and the default name is sites.xml.
>Urlcache_Filename sites.xml

### Enable quick access to Urlcache
* Does not affect the default access method of UrlCache.
* Values are "1" (on) and "0" (off). The default is 0 to close.
>Urlcache_Request_Easy 0


## [Task]
### Timeer configuration--storage time for automatic backup of DB data
* The unit is in minutes. The default is 1.
* Warning: If there is a lot of data in the database, when the service is restarted, if the time to reload the data into the memory exceeds the set backup time, the data entity file will be overwritten, which is a very serious data loss error.  
>Task_DataBackup 1

### Timeer configuration--automatic backup of the storage time of user authentication data
* The unit is in minutes. The default is 1.
* Warning: If there is a lot of data in the database, when the service is restarted, if the time to reload the data into the memory exceeds the set backup time, the data entity file will be overwritten, which is a very serious data loss error.  
>Task_DataAuthBackup 1

### Timer configuration--the time to automatically clean up expired data in the database
* The unit is in minutes. The default is 1.
>Task_ClearExpireData 1

### Timer configuration--time to automatically back up server cluster information data
* The unit is in minutes. The default is 1.
* Warning: If there is a lot of data in the database, when the service is restarted, if the time to reload the data into the memory exceeds the set backup time, the data entity file will be overwritten, which is a very serious data loss error.  
>Task_ServerListBackup 1

### Timer configuration--time to automatically back up Gossip service cluster information
* The unit is in minutes. The default is 1.
* Warning: If there is a lot of data in the database, when the service is restarted, if the time to reload the data into the memory exceeds the set backup time, the data entity file will be overwritten, which is a very serious data loss error.
>Task_ServerGossipListBackup 1

### Timer configuration--automatically output Owl memory information
* The unit is in minutes. The default is 1. 
>Task_MemoryInfoToLog 1

### Enable Task execution success log output. A value of 1 prints logs, and any other value masks logs.  
>Task_OutputSuccessLog 1

doc/zh/2.config.md
[<p align="right">->返回目录</p>](0.directory.md)

# owlcache 配置文件
* 程序启动时，您动态配置的参数会覆盖掉*.conf文件中的参数

## 配置文件书写约定
>1.以#开头表示这一行是注释内容  
>2.以[]表示分组信息(更直观的展示配置信息)  
>3.配置参数名与参数值以一个空格间隔。  

## [Owlcache]
### 设置程序监听的IP
* 如果您正常启动后无法访问服务，请检查下防火墙之类的软件是否限制了您的访问
>Host 0.0.0.0

### 设置程序响应IP
* 默认本机
* 在TCP、HTTP返回的响应结果Json字符串中使用，注意局域网、外网环境中实际部署时IP的正确性!
>ResponseHost 127.0.0.1

### 日志文件路径
* 默认日志会生成到服务运行目录
* !!!如果在命令行中启动服务时指定了Logfile值，而配置文件这里没有注释掉则会以配置文件为准
>Logfile ./log_file/

### 数据库文件路径
* 默认数据库文件会生成到服务运行目录
>DBfile ./db_file/


## [TCP]
### TCP监听端口
* 默认为7720
* 如果不填写或者端口值为0 则默认监听7720端口
>Tcpport 7720

### 是否开启Tcp服务
* TCP模式下无密码认证(生产环境部署请根据实际情况屏蔽TCP的访问端口等等，注意使用安全)
* 值为"1"(开启服务)和"0"(关闭服务)。默认为1开启服务。
>CloseTcp 1


## [HTTP]
### HTTP监听端口
* 默认为7721。
* 如果不填写或者端口值为0 则默认监听7721端口.
>Httpport 7721

### HTTP密码
* 强制且必须的，不要设置的过于简单
* 请设置密码。密码要求长度不能低于十位数，不能是纯数字，建议采用大写字母+小写字母+数字+简单符号
>Pass 

### Tonken过期时间
* 为"Pass"命令产生的Tonken值设置一个过期时间。
* 默认为0。单位是秒。如果值为0，则永不过期。
* 警告:设置永不过期，也许会存在安全风险。
>Tonken_expire_time 0

### 设置Tonken值的密钥
* 建议更改默认值。
>Tonken_jwt_secret hello_owlcache

### HTTPS相关配置
* Open_Https 是否开启HTTPS。值为0(关闭)、1(开启)。默认关闭。
* Https_CertFile Cert文件路径。例如:/www/server.crt、D:/key/server.crt。
* Https_KeyFile Key文件路径。例如:/www/server.key、D:/key/server.key。
* HttpsClient_InsecureSkipVerify 当开启HTTPS模式后，owlcache之间互相通讯时是否校验证书合法性。值为0(关闭)、1(开启)。默认关闭。开启时不会校验证书合法性。
>Open_Https 0

>Https_CertFile /www/server.crt

>Https_KeyFile /www/server.key

>HttpsClient_InsecureSkipVerify 0

### 跨域相关配置
* "跨域资源共享"(Cross-origin resource sharing)配置。
* Cors值为"1"(开启服务)和"0"(关闭服务)。默认为0关闭服务不允许跨域。
* Access_Control_Allow_Origin值默认为"*"(允许所有域的请求)。您可以将此值设置为您指定的域。
* 注意:谨慎设置小心CSRF攻击。
>Cors 0

>Access_Control_Allow_Origin *


## [HttpGroupData]
### 集群互相通信时的请求超时时间
* 单位毫秒。默认2700毫秒超时。实际上内网环境的数据互通时间很低,可以根据实际情况来设置。
>HttpClientRequestTimeout 2700

### Http客户端请求设置
* 请求得到得数据将会短时间缓存在本地,避免同一个Key在高并发状态下反复对其他节点疯狂查询。适当的延迟查询有助于性能和效率得优化。
* 单位毫秒，默认这个缓存生命周期为5000毫秒。值为0则不进行缓存(适合并发量小，数据实时性要求高的场景)。
>HttpClientRequestLocalCacheLifeTime 5000


## [GroupDataSync]

### 是否开启集群数据同步
* 0表示不开启。1表示开启。默认不开启。
>GroupDataSync 0

### 设置Gossip监听端口
* 启用集群数据同步服务该项才会生效。
* 集群数据同步服务依赖于Gossip协议。建议不设置该项,系统会自动进行分配端口。默认值为0(系统自动监听一个端口并在启动信息输出该端口)。
>Gossipport 0

### 整个Gossip集群需要统一的令牌
* 启用集群数据同步服务该项才会生效。
* 集群中通过Gossip协议交换数据的令牌。默认为空，你需要为集群设置一个相同密码。
>GossipDataSyncAuthKey 

### 集群同步数据互相通信时的请求超时时间
* 单位毫秒。默认5000毫秒超时。实际上内网环境的数据互通时间很低,可以根据实际情况来设置。
>GossipHttpClientRequestTimeout 5000


## [Memcache_Client]
### 开启从memcache导入数据
* 是否开启查询不存在的数据时,从memcache查询并存入本地数据库。0表示不开启。
>Get_data_from_memcache 0

### 查询的memcache列表
* 不同节点之间需要使用"|"符号间隔。例如:127.0.0.1:11211|192.168.0.77:11211。默认为空值。
>Memcache_list 

### 导入数据的过期时间
* 为从memcache存入本地数据库的Key设置一个过期时间。默认为0，永久不过期。单位是秒。
>Get_memcache_data_set_expire_time 0

### MemcacheClient客户端请求设置
* Max_Error_Number超过限定值时，MemcacheClient请求将“暂停”Sleeptime值，来优化程序响应速度。
* Sleeptime代表睡眠时间，单位秒，默认最低睡眠2秒。
* Max_Error_Number代表最大失败请求数，单位是整数，默认最小错误请求2次。
>MemcacheClient_Request_Timeout_Sleeptime 2

>MemcacheClient_Request_Max_Error_Number 2


## [Redis_Client]
### 开启从Redis导入数据
* 是否开启查询不存在的数据时,从redis查询并存入本地数据库。0表示不开启。
>Get_data_from_redis 0

### 需要查询的Redis地址
* 不能是空值。  
>Redis_Addr 

### 需要查询的Redis密码
* 不能是空值。  
>Redis_Password 

### 需要查询Redis的数据库
* 默认为0
>Redis_DB 0

### 导入数据的过期时间
* 为从Redis存入本地数据库的Key设置一个过期时间。默认为0，永久不过期。单位是秒。
>Get_redis_data_set_expire_time 0

### RedisClient客户端请求设置
* RedisClient客户端请求设置。Max_Error_Number超过限定值时，RedisClient请求将“暂停”Sleeptime值，来优化程序响应速度。
* Sleeptime代表睡眠时间，单位秒，默认最低睡眠2秒。
* Max_Error_Number代表最大失败请求数，单位是整数，默认最小错误请求2次。
>RedisClient_Request_Timeout_Sleeptime 2

>RedisClient_Request_Max_Error_Number 2


## [Url_Cache]
### 开启从URL地址导入数据
* URL代理访问后将得到的HTTP响应数据缓存到Owlcache中
* 值为"1"(开启服务)和"0"(关闭服务)。默认为0关闭服务。
>Open_Urlcache 0

### 开启URL缓存后的需要加载的配置文件的名称
* 格式为XML，它存在于DBfile配置项目录中，默认名称为sites.xml。
>Urlcache_Filename sites.xml

### 开启Urlcache的快捷访问
* 不影响UrlCache的默认访问方式。
* 值为"1"(开启)和"0"(关闭)。默认为0关闭。
>Urlcache_Request_Easy 0


## [Task]
### 定时配置-自动备份DB数据的存储时间
* 单位分钟。默认为1.
* 警告:如果数据库中数据非常多,重新启动服务时,假如数据重新装载到内存的时间，超过了设置的备份时间就会导致数据实体文件被覆盖掉，这是非常严重的数据丢失错误。  
>Task_DataBackup 1

### 定时配置-自动备份用户认证数据的存储时间
* 单位分钟。默认为1.
* 警告:如果数据库中数据非常多,重新启动服务时,假如数据重新装载到内存的时间，超过了设置的备份时间就会导致数据实体文件被覆盖掉，这是非常严重的数据丢失错误。  
>Task_DataAuthBackup 1

### 定时配置-自动清理数据库中过期数据的时间
* 单位分钟。默认为1.
>Task_ClearExpireData 1

### 定时配置-自动备份服务器集群信息数据的时间
* 单位分钟。默认为1.
* 警告:如果数据库中数据非常多,重新启动服务时,假如数据重新装载到内存的时间，超过了设置的备份时间就会导致数据实体文件被覆盖掉，这是非常严重的数据丢失错误。  
>Task_ServerListBackup 1

### 定时配置-自动备份Gossip服务集群信息的时间
* 单位分钟。默认为1.
* 警告:如果数据库中数据非常多,重新启动服务时,假如数据重新装载到内存的时间，超过了设置的备份时间就会导致数据实体文件被覆盖掉，这是非常严重的数据丢失错误。  
>Task_ServerGossipListBackup 1

### 定时配置-自动输出Owl的内存信息
* 单位分钟。默认为1.  
>Task_MemoryInfoToLog 1

### 开启Task执行成功日志输出。值为1，输出日志，其他任何值则会屏蔽日志。  
>Task_OutputSuccessLog 1

config/config.go
package config

import (
	"bufio"
	"fmt"
	"os"
	"strings"

	"github.com/xssed/owlcache/tools"
)

//创建一个全局配置变量
var OwlConfigModel *OwlConfig

//配置文件模型
type OwlConfig struct {
	Configfile                               string //配置文件路径
	Logfile                                  string //日志文件路径
	DBfile                                   string //数据库文件路径
	Pass                                     string //owlcache密钥
	Tonken_expire_time                       string //为Pass命令产生的Tonken值设置一个过期时间
	Tonken_jwt_secret                        string //设置Tonken值的密钥
	Host                                     string //主机地址
	ResponseHost                             string //程序响应IP,在TCP、HTTP返回的响应结果Json字符串中使用
	Tcpport                                  string //Tcp监听端口
	Httpport                                 string //Http监听端口
	HttpClientRequestTimeout                 string //集群互相通信时的请求超时时间
	HttpClientRequestLocalCacheLifeTime      string //Http客户端请求优化设置。请求得到得数据将会短时间缓存在本地,避免同一个Key在高并发状态下反复对其他节点疯狂查询。适当的延迟查询有助于性能和效率得优化。单位毫秒，默认这个缓存生命周期为5000毫秒。
	GroupDataSync                            string //是否开启集群数据同步。0表示不开启。1表示开启。默认不开启。
	Gossipport                               string //启用Gossip服务该项才会生效。Gossip监听端口，默认值为0(系统自动监听一个端口并在启动信息输出该端口)。
	GossipDataSyncAuthKey                    string //启用Gossip服务该项才会生效。集群中通过Gossip协议交换数据的令牌。整个集群需要统一的令牌。
	GossipHttpClientRequestTimeout           string //集群同步数据互相通信时的请求超时时间
	Task_DataBackup                          string //自动备份DB数据的存储时间
	Task_DataAuthBackup                      string //自动备份用户认证数据的存储时间
	Task_ClearExpireData                     string //自动清理数据库中过期数据的时间
	Task_ServerListBackup                    string //自动备份服务器集群信息数据的时间
	Task_ServerGossipListBackup              string //自动备份Gossip集群信息数据的时间
	Task_MemoryInfoToLog                     string //自动输出Owl的内存信息的时间
	Task_OutputSuccessLog                    string //开启Task执行成功日志输出
	Get_data_from_memcache                   string //是否开启查询不存在的数据时,从memcache查询并存入本地数据库
	Memcache_list                            string //需要查询的memcache列表,不同节点之间需要使用"|"符号间隔。
	Get_memcache_data_set_expire_time        string //为从memcache存入本地数据库的Key设置一个过期时间。默认为0，永久不过期。单位是秒。
	Get_data_from_redis                      string //是否开启查询不存在的数据时,从redis查询并存入本地数据库。0表示不开启。
	Redis_Addr                               string //需要查询的Redis地址。
	Redis_Password                           string //需要查询的Redis密码。不能是空值。
	Redis_DB                                 string //需要查询Redis的数据库。默认为0。
	Get_redis_data_set_expire_time           string //为从Redis存入本地数据库的Key设置一个过期时间。默认为0，永久不过期。单位是秒。
	Open_Https                               string //是否开启HTTPS。值为0(关闭)、1(开启)。默认关闭。
	Https_CertFile                           string //Cert文件路径。
	Https_KeyFile                            string //Key文件路径。
	HttpsClient_InsecureSkipVerify           string //当开启HTTPS模式后，owlcache之间互相通讯时是否校验证书。值为0(关闭)、1(开启)。默认关闭。开启时不会校验证书合法性。
	Open_Websocket_Server                    string //是否开启Websocket Server。值为0(关闭)、1(开启)。默认关闭。
	GroupData_Mode                           string //owlcache集群方式，分为“Http”(短链接)和“Websocket”(长链接)，两个选项，默认是“Http”短链接。
	Websocket_Client_WriteWait               string //客户端写超时。单位秒。默认10秒。
	Websocket_Client_MaxMessageSize          string //客户端支持接受的消息最大长度，单位字节。默认7000000字节。
	Websocket_Client_MinRecTime              string //客户端断开与服务端后最小重连时间间隔。单位秒。默认2秒。
	Websocket_Client_MaxRecTime              string //客户端断开与服务端后最大重连时间间隔。单位秒。默认60秒。
	Websocket_Client_MessageBufferSize       string //客户端消息发送缓冲池大小，单位字节。默认1024字节。
	CloseTcp                                 string //是否关闭Tcp服务(因为TCP模式下无密码认证)  值为"1"(开启)和"0"(关闭)。默认为1开启服务。
	Cors                                     string //是否开启跨域
	Access_Control_Allow_Origin              string //设置指定的域
	MemcacheClient_Request_Timeout_Sleeptime string //MemcacheClient客户端请求设置。请求的睡眠时间。单位是整数。
	MemcacheClient_Request_Max_Error_Number  string //MemcacheClient客户端Max_Error_Number超过限定值时，MemcacheClient请求将“暂停”Sleeptime值，来优化程序响应速度。
	RedisClient_Request_Timeout_Sleeptime    string //RedisClient客户端请求设置。请求的睡眠时间。单位是整数。
	RedisClient_Request_Max_Error_Number     string //RedisClient客户端Max_Error_Number超过限定值时，RedisClient请求将“暂停”Sleeptime值，来优化程序响应速度。
	Open_Urlcache                            string //URL代理访问后将得到的HTTP响应数据缓存到Owlcache中，值为"1"(开启服务)和"0"(关闭服务)。默认为0关闭服务。
	Urlcache_Filename                        string //开启URL缓存后的需要加载的配置文件的名称，格式为XML，它存在于DBfile配置项目录中，默认名称为sites.xml。
	Urlcache_Request_Easy                    string //开启Urlcache的快捷访问。不影响UrlCache的默认访问方式。值为"1"(开启)和"0"(关闭)。默认为0关闭。
}

//创建一个默认配置文件的实体
func NewDefaultOwlConfig() *OwlConfig {
	return &OwlConfig{
		Configfile:                               "owlcache.conf",
		Logfile:                                  "",
		DBfile:                                   "",
		Pass:                                     "",
		Tonken_expire_time:                       "0",
		Tonken_jwt_secret:                        "hello_owlcache",
		Host:                                     "0.0.0.0",
		ResponseHost:                             "127.0.0.1",
		Tcpport:                                  "7720",
		Httpport:                                 "7721",
		HttpClientRequestTimeout:                 "2700",
		HttpClientRequestLocalCacheLifeTime:      "5000",
		GroupDataSync:                            "0",
		Gossipport:                               "0",
		GossipDataSyncAuthKey:                    "",
		GossipHttpClientRequestTimeout:           "5000",
		Task_DataBackup:                          "1",
		Task_DataAuthBackup:                      "1",
		Task_ClearExpireData:                     "1",
		Task_ServerListBackup:                    "1",
		Task_ServerGossipListBackup:              "1",
		Task_MemoryInfoToLog:                     "1",
		Task_OutputSuccessLog:                    "1",
		Get_data_from_memcache:                   "0",
		Memcache_list:                            "",
		Get_memcache_data_set_expire_time:        "0",
		Get_data_from_redis:                      "0",
		Redis_Addr:                               "",
		Redis_Password:                           "",
		Redis_DB:                                 "0",
		Get_redis_data_set_expire_time:           "0",
		Open_Https:                               "0",
		Https_CertFile:                           "/www/server.crt",
		Https_KeyFile:                            "/www/server.key",
		HttpsClient_InsecureSkipVerify:           "0",
		Open_Websocket_Server:                    "0",
		GroupData_Mode:                           "Http",
		Websocket_Client_WriteWait:               "10",
		Websocket_Client_MaxMessageSize:          "7000000",
		Websocket_Client_MinRecTime:              "2",
		Websocket_Client_MaxRecTime:              "60",
		Websocket_Client_MessageBufferSize:       "1024",
		CloseTcp:                                 "1",
		Cors:                                     "0",
		Access_Control_Allow_Origin:              "*",
		MemcacheClient_Request_Timeout_Sleeptime: "2",
		MemcacheClient_Request_Max_Error_Number:  "2",
		RedisClient_Request_Timeout_Sleeptime:    "2",
		RedisClient_Request_Max_Error_Number:     "2",
		Open_Urlcache:                            "0",
		Urlcache_Filename:                        "sites.xml",
		Urlcache_Request_Easy:                    "0",
	}
}

//缓存系统初始化加载配置
func ConfigInit() {

	//读取配置文件获取一个最终的配置模型
	var config = make(map[string]string)

	//创建一个默认初始化配置模型
	OwlConfigModel = NewDefaultOwlConfig()

	//打开配置文件
	config_file, err := os.Open(OwlConfigModel.Configfile)
	defer config_file.Close()
	//监控错误
	if err != nil {
		fmt.Println(err)
		fmt.Print("Can not read configuration file. now exit\n")
		os.Exit(0)
	}
	//判断文件是否为UTF-8编码
	if !tools.ValidUTF8(OwlConfigModel.Configfile) {
		fmt.Print("The configuration file is not UTF-8 encode. now exit\n")
		os.Exit(0)
	}
	//将内容读入缓冲区
	buff := bufio.NewReader(config_file)
	//读取配置文件
	for {
		line, _, err := buff.ReadLine()
		if err != nil {
			break
		}
		rs := []rune(string(line))
		if string(rs[0:1]) == `#` || len(string(line)) < 3 {
			continue
		}
		if string(rs[0:1]) == `[` || len(string(line)) < 3 {
			continue
		}
		config_string_arr := strings.Split(string(line), " ")

		type_name := config_string_arr[0]
		type_value := config_string_arr[1]

		config[type_name] = type_value
	}

	//fmt.Println(OwlConfigModel) //打印出默认配置信息
	//fmt.Println(config)         //打印出*.conf中的配置信息
	//将文本配置绑定到全局配置
	ConfigBind(config, OwlConfigModel)
	//最后检查参数
	OwlConfigModel = CmdParamInit(OwlConfigModel)
	//fmt.Println(OwlConfigModel) //打印出最终赋值后的配置信息

	//执行步骤信息
	fmt.Println("owlcache  system configuration initialization is complete...")

	//检查是否启动URL缓存功能
	if OwlConfigModel.Open_Urlcache == "1" {
		//URL缓存初始化加载配置
		OwlUCConfigInit()
	}

}

//将文本配置绑定到全局配置
func ConfigBind(config map[string]string, param *OwlConfig) {

	if len(config["Host"]) > 3 {
		//!!!如果在命令行中启动服务时指定了Host值，而配置文件这里没有注释掉则会以配置文件为准
		param.Host = config["Host"]
	}
	if len(config["ResponseHost"]) > 3 {
		param.ResponseHost = config["ResponseHost"]
	}
	if len(config["Tcpport"]) > 1 {
		param.Tcpport = config["Tcpport"]
	}
	if len(config["Httpport"]) > 1 {
		param.Httpport = config["Httpport"]
	}
	if len(config["Pass"]) > 1 {
		param.Pass = config["Pass"]
	}
	if len(config["Tonken_expire_time"]) > 1 {
		param.Tonken_expire_time = config["Tonken_expire_time"]
	}
	if len(config["Tonken_jwt_secret"]) > 1 {
		param.Tonken_jwt_secret = config["Tonken_jwt_secret"]
	}
	if len(config["Logfile"]) > 3 {
		//!!!如果在命令行中启动服务时指定了Logfile值，而配置文件这里没有注释掉则会以配置文件为准
		param.Logfile = config["Logfile"]
	}
	if len(config["DBfile"]) > 3 {
		param.DBfile = config["DBfile"]
	}
	if len(config["HttpClientRequestTimeout"]) >= 1 {
		param.HttpClientRequestTimeout = config["HttpClientRequestTimeout"]
	}
	if len(config["HttpClientRequestLocalCacheLifeTime"]) >= 1 {
		param.HttpClientRequestLocalCacheLifeTime = config["HttpClientRequestLocalCacheLifeTime"]
	}
	if len(config["GroupDataSync"]) >= 1 {
		param.GroupDataSync = config["GroupDataSync"]
	}
	if len(config["Gossipport"]) >= 1 {
		param.Gossipport = config["Gossipport"]
	}
	if len(config["GossipDataSyncAuthKey"]) >= 1 {
		param.GossipDataSyncAuthKey = config["GossipDataSyncAuthKey"]
	}
	if len(config["GossipHttpClientRequestTimeout"]) >= 1 {
		param.GossipHttpClientRequestTimeout = config["GossipHttpClientRequestTimeout"]
	}
	if len(config["Task_DataBackup"]) >= 1 {
		param.Task_DataBackup = config["Task_DataBackup"]
	}
	if len(config["Task_DataAuthBackup"]) >= 1 {
		param.Task_DataAuthBackup = config["Task_DataAuthBackup"]
	}
	if len(config["Task_ClearExpireData"]) >= 1 {
		param.Task_ClearExpireData = config["Task_ClearExpireData"]
	}
	if len(config["Task_ServerListBackup"]) >= 1 {
		param.Task_ServerListBackup = config["Task_ServerListBackup"]
	}
	if len(config["Task_ServerGossipListBackup"]) >= 1 {
		param.Task_ServerGossipListBackup = config["Task_ServerGossipListBackup"]
	}
	if len(config["Task_MemoryInfoToLog"]) >= 1 {
		param.Task_MemoryInfoToLog = config["Task_MemoryInfoToLog"]
	}
	if len(config["Task_OutputSuccessLog"]) >= 1 {
		param.Task_OutputSuccessLog = config["Task_OutputSuccessLog"]
	}
	if len(config["Get_data_from_memcache"]) >= 1 {
		param.Get_data_from_memcache = config["Get_data_from_memcache"]
	}
	if len(config["Memcache_list"]) >= 1 {
		param.Memcache_list = config["Memcache_list"]
	}
	if len(config["Get_memcache_data_set_expire_time"]) >= 1 {
		param.Get_memcache_data_set_expire_time = config["Get_memcache_data_set_expire_time"]
	}
	if len(config["Get_data_from_redis"]) >= 1 {
		param.Get_data_from_redis = config["Get_data_from_redis"]
	}
	if len(config["Redis_Addr"]) >= 1 {
		param.Redis_Addr = config["Redis_Addr"]
	}
	if len(config["Redis_Password"]) >= 1 {
		param.Redis_Password = config["Redis_Password"]
	}
	if len(config["Redis_DB"]) >= 1 {
		param.Redis_DB = config["Redis_DB"]
	}
	if len(config["Get_redis_data_set_expire_time"]) >= 1 {
		param.Get_redis_data_set_expire_time = config["Get_redis_data_set_expire_time"]
	}
	if len(config["Open_Https"]) >= 1 {
		param.Open_Https = config["Open_Https"]
	}
	if len(config["Https_CertFile"]) >= 1 {
		param.Https_CertFile = config["Https_CertFile"]
	}
	if len(config["Https_KeyFile"]) >= 1 {
		param.Https_KeyFile = config["Https_KeyFile"]
	}
	if len(config["HttpsClient_InsecureSkipVerify"]) >= 1 {
		param.HttpsClient_InsecureSkipVerify = config["HttpsClient_InsecureSkipVerify"]
	}
	if len(config["Open_Websocket_Server"]) >= 1 {
		param.Open_Websocket_Server = config["Open_Websocket_Server"]
	}
	if len(config["GroupData_Mode"]) >= 1 {
		param.GroupData_Mode = config["GroupData_Mode"]
	}
	if len(config["Websocket_Client_WriteWait"]) >= 1 {
		param.Websocket_Client_WriteWait = config["Websocket_Client_WriteWait"]
	}
	if len(config["Websocket_Client_MaxMessageSize"]) >= 1 {
		param.Websocket_Client_MaxMessageSize = config["Websocket_Client_MaxMessageSize"]
	}
	if len(config["Websocket_Client_MinRecTime"]) >= 1 {
		param.Websocket_Client_MinRecTime = config["Websocket_Client_MinRecTime"]
	}
	if len(config["Websocket_Client_MaxRecTime"]) >= 1 {
		param.Websocket_Client_MaxRecTime = config["Websocket_Client_MaxRecTime"]
	}
	if len(config["Websocket_Client_MessageBufferSize"]) >= 1 {
		param.Websocket_Client_MessageBufferSize = config["Websocket_Client_MessageBufferSize"]
	}
	if len(config["CloseTcp"]) >= 1 {
		param.CloseTcp = config["CloseTcp"]
	}
	if len(config["Cors"]) >= 1 {
		param.Cors = config["Cors"]
	}
	if len(config["Access_Control_Allow_Origin"]) >= 1 {
		param.Access_Control_Allow_Origin = config["Access_Control_Allow_Origin"]
	}
	if len(config["MemcacheClient_Request_Timeout_Sleeptime"]) >= 1 {
		param.MemcacheClient_Request_Timeout_Sleeptime = config["MemcacheClient_Request_Timeout_Sleeptime"]
	}
	if len(config["MemcacheClient_Request_Max_Error_Number"]) >= 1 {
		param.MemcacheClient_Request_Max_Error_Number = config["MemcacheClient_Request_Max_Error_Number"]
	}
	if len(config["RedisClient_Request_Timeout_Sleeptime"]) >= 1 {
		param.RedisClient_Request_Timeout_Sleeptime = config["RedisClient_Request_Timeout_Sleeptime"]
	}
	if len(config["RedisClient_Request_Max_Error_Number"]) >= 1 {
		param.RedisClient_Request_Max_Error_Number = config["RedisClient_Request_Max_Error_Number"]
	}
	if len(config["Open_Urlcache"]) >= 1 {
		param.Open_Urlcache = config["Open_Urlcache"]
	}
	if len(config["Urlcache_Filename"]) >= 1 {
		param.Urlcache_Filename = config["Urlcache_Filename"]
	}
	if len(config["Urlcache_Request_Easy"]) >= 1 {
		param.Urlcache_Request_Easy = config["Urlcache_Request_Easy"]
	}

}
